style: Switch off eslint rule linebreak-style

sudo-suhas opened the pull request:
The rule conflicts with git `core.autocrlf`, enabled through `.gitattributes`, on windows.

**Before**
```
λ yarn run lint
yarn run v1.2.0
$ eslint ./src ./test ./build

E:\Projects\repos\depcheck\src\check.js
    1:21  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    2:25  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    3:29  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    4:31  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    5:35  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    6:46  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    7:55  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    8:36  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    9:1   error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   10:25  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   11:8   error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   12:49  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
....

✖ 3351 problems (3351 errors, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**After**
```
λ yarn run lint
yarn run v1.2.0
$ eslint ./src ./test ./build
Done in 7.64s.
```

lijunle commented:
What value are you set your global rule `core.autocrlf` to? I worked on Windows for a long time and all settings are good to me.

sudo-suhas commented:
```
λ git config --global core.autocrlf
false
```
I think the `.gitattributes` file takes precedence. I tried a fresh clone:

```
λ git clone https://github.com/depcheck/depcheck.git
Cloning into 'depcheck'...
remote: Counting objects: 3098, done.
remote: Total 3098 (delta 0), reused 0 (delta 0), pack-reused 3098
Receiving objects: 100% (3098/3098), 456.37 KiB | 264.00 KiB/s, done.
Resolving deltas: 100% (1721/1721), done.

λ cd depcheck

λ yarn -s

λ yarn run lint
yarn run v1.2.0
warning ..\package.json: No license field
$ eslint ./src ./test ./build

E:\Projects\experiments\depcheck\src\check.js
    1:21  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    2:25  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    3:29  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    4:31  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    5:35  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    6:46  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    7:55  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    8:36  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
    9:1   error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   10:25  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   11:8   error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   12:49  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   13:17  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   14:20  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
   15:18  error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style
....

✖ 3481 problems (3481 errors, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

lijunle commented:
Not sure why you are getting bad line feed. But, you could fix it with one command `npm run lint -- --fix` which run [`eslint --fix`](https://eslint.org/docs/user-guide/migrating-from-jscs#fix) under the hook.

sudo-suhas commented:
This is the setting responsible for it:

https://github.com/depcheck/depcheck/blob/153aa50925e4a02f1b418a7d1349bff8f5f925eb/.gitattributes#L1

Read more here - https://git-scm.com/docs/gitattributes, https://stackoverflow.com/questions/170961/whats-the-best-crlf-carriage-return-line-feed-handling-strategy-with-git, https://stackoverflow.com/questions/3206843/how-line-ending-conversions-work-with-git-core-autocrlf-between-different-operat

> This attribute enables and controls end-of-line normalization. When a text file is normalized, its line endings are converted to LF in the repository.

From what I understand, this is still the best strategy to keep CRLF out of the git repo. However, it conflicts with the eslint rule and hence this PR.

If I run `npm run lint -- --fix`, it ends up modifying all files:

<details>

<summary> git status </summary>

```
λ git status
On branch master
Your branch is up-to-date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   build/check-json.js
        modified:   build/component.js
        modified:   src/check.js
        modified:   src/cli.js
        modified:   src/constants.js
        modified:   src/detector/expressViewEngine.js
        modified:   src/detector/gruntLoadTaskCallExpression.js
        modified:   src/detector/importDeclaration.js
        modified:   src/detector/requireCallExpression.js
        modified:   src/detector/requireResolveCallExpression.js
        modified:   src/index.js
        modified:   src/parser/coffee.js
        modified:   src/parser/es6.js
        modified:   src/parser/es7.js
        modified:   src/parser/jsx.js
        modified:   src/parser/sass.js
        modified:   src/parser/typescript.js
        modified:   src/special/babel.js
        modified:   src/special/bin.js
        modified:   src/special/commitizen.js
        modified:   src/special/eslint.js
        modified:   src/special/feross-standard.js
        modified:   src/special/gulp-load-plugins.js
        modified:   src/special/mocha.js
        modified:   src/special/webpack.js
        modified:   src/utils/get-scripts.js
        modified:   src/utils/index.js
        modified:   test/cli.js
        modified:   test/fake_detectors/dependCallExpression.js
        modified:   test/fake_detectors/exception.js
        modified:   test/fake_modules/bad/index.js
        modified:   test/fake_modules/bad_deep/test/sandbox/index.js
        modified:   test/fake_modules/bad_es6/index.js
        modified:   test/fake_modules/bin_js/index.js
        modified:   test/fake_modules/depend/index.js
        modified:   test/fake_modules/dev/index.js
        modified:   test/fake_modules/empty_dep/index.js
        modified:   test/fake_modules/express_view_engine/index.js
        modified:   test/fake_modules/good/index.js
        modified:   test/fake_modules/good_es6/index.js
        modified:   test/fake_modules/good_es7/index.js
        modified:   test/fake_modules/grunt-tasks/index.js
        modified:   test/fake_modules/grunt/index.js
        modified:   test/fake_modules/ignore_number/index.js
        modified:   test/fake_modules/jsx_js/index.js
        modified:   test/fake_modules/missing/index.js
        modified:   test/fake_modules/missing_ignore/index.js
        modified:   test/fake_modules/missing_nested/index.js
        modified:   test/fake_modules/missing_nested/nested/index.js
        modified:   test/fake_modules/missing_peer_deps/index.js
        modified:   test/fake_modules/nested/index.js
        modified:   test/fake_modules/optional_dep/index.js
        modified:   test/fake_modules/peer_dep/index.js
        modified:   test/fake_modules/peer_dep_nested/nested/index.js
        modified:   test/fake_modules/require_dynamic/index.js
        modified:   test/fake_modules/require_nothing/index.js
        modified:   test/fake_modules/require_resolve/index.js
        modified:   test/fake_modules/require_resolve_missing/index.js
        modified:   test/fake_modules/scoped_module/index.js
        modified:   test/fake_modules/shebang/index.js
        modified:   test/fake_modules/unreadable_deep/deep/nested/index.js
        modified:   test/fake_parsers/exception.js
        modified:   test/fake_parsers/importList.js
        modified:   test/fake_parsers/multiple.js
        modified:   test/index.js
        modified:   test/spec.js
        modified:   test/special/babel.js
        modified:   test/special/bin.js
        modified:   test/special/commitizen.js
        modified:   test/special/eslint.js
        modified:   test/special/feross-standard.js
        modified:   test/special/gulp-load-plugins.js
        modified:   test/special/mocha.js
        modified:   test/special/node_modules/binary-no-package/index.js
        modified:   test/special/node_modules/eslint-config-airbnb/base.js
        modified:   test/special/node_modules/eslint-config-airbnb/index.js
        modified:   test/special/node_modules/eslint-config-airbnb/react.js
        modified:   test/special/node_modules/eslint-config-preset/index.js
        modified:   test/special/node_modules/webpack/index.js
        modified:   test/special/webpack.js
        modified:   test/utils.js
```

</details>

lijunle commented:
> This attribute enables and controls end-of-line normalization. When a text file is normalized, its line endings are converted to LF in the repository.
> From what I understand, this is still the best strategy to keep CRLF out of the git repo. However, it conflicts with the eslint rule and hence this PR.

The `.gitattribute` file is only controlling the CRLF inside the repo, not in the working directory.

> If I run npm run lint -- --fix, it ends up modifying all files.

After you run `git add .`, do they go away?

BTW, which version of Git are you using?

sudo-suhas commented:
> The `.gitattribute` file is only controlling the CRLF inside the repo, not in the working directory.

I am not sure what you mean by this. Could you please explain?

> After you run `git add .`, do they go away?

Yes they did. However, This is still not a good experience for a new contributor on windows. Either we can switch of the eslint rule or tweak the `.gitattributes` for js files as done here - https://github.com/kentcdodds/cross-env/blob/36b009e25ab5143fbe5d7e8ec92983dc700c60ac/.gitattributes

lijunle commented:
The .gitattributes is *not* used to control how CRLF is check *out*. Change
it had no sense.

It is the *core.autocrlf* config to control that. What is your value is
this config?

2017年10月18日 10:21，"Suhas Karanth" <notifications@github.com>写道：

The .gitattribute file is only controlling the CRLF inside the repo, not in
the working directory.

I am not sure what you mean by this. Could you please explain?

After you run git add ., do they go away?

Yes they did. However, This is still not a good experience for a new
contributor on windows. Either we can switch of the eslint rule or tweak
the .gitattributes for js files as done here -
https://github.com/kentcdodds/cross-env/blob/36b009e25ab5143fbe5d7e8ec92983
dc700c60ac/.gitattributes

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/depcheck/depcheck/pull/228#issuecomment-337441119>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABPIdGrGjFim5f0ffO8VwyePylbefLjNks5stWCjgaJpZM4P5nhn>
.

sudo-suhas commented:
I had mentioned this before, I have `core.autocrlf` set to `false`. I changed it to `true` and tried a fresh clone and I still get the errors. I am not sure how I can convince you but the `.gitattributes` file takes precedence and it is responsible for normalization of line endings. See [kentcdodds/cross-env#47](https://github.com/kentcdodds/cross-env/issues/47)

lijunle commented:
> I had mentioned this before, I have core.autocrlf set to false. I changed it to true and tried a fresh clone and I still get the errors.

Could you just remove the `core.autocrlf` entry, globally.

> I am not sure how I can convince you but the `.gitattributes` file takes precedence and it is responsible for normalization of line endings. 

It is not working like that. When we say *normalization*, it happens on people check in the files into the repo, it does not happen on files are checked out from the repo. If we change normalize the line feed to CRLF, the repo contains the CRLF, when check out to Linux/Mac environment, they will get CRLF line feed. If you check the doc, there is no way to convert the CRLF back to LF in Linux/Mac when check out. If most of the people working on project in using Windows, it is fine to do it. But it is not good for a open source project.

sudo-suhas commented:
So here's the thing. I have tried a fresh clone with both possible settings for `core.autocrlf` and got errors for both. I have shared links that explain how line endings work with `text=auto`(here's another - https://help.github.com/articles/dealing-with-line-endings/). I am fairly sure that I have tried my level best to explain what's going on. Let me know if you change your mind. For now, I am giving up.
